### PR TITLE
V5 deserialization tests

### PR DESCRIPTION
**Pull request by OoLunar**
# Summary
Utilizing https://github.com/discord-payloads/discord-payloads, we can test our gateway entities by deserializing the payloads provided. We did obtain permission to do so from @Victorsitou.

# Changes proposed
- Added https://github.com/discord-payloads/discord-payloads as a submodule
- Added `DSharpPlus.Test`
- Allows `dotnet test` to be utilized on new commits.

### CODE REVIEW

File: DSharpPlus.Core/Attributes/DiscordGatewayEventName.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+using System;
+
+namespace DSharpPlus.Core.Attributes
+{
+    [AttributeUsage(AttributeTargets.Class, Inherited = false, AllowMultiple = true)]
+    public sealed class DiscordGatewayEventNameAttribute : Attribute
```
Comment (ExaInsanity) [inline DSharpPlus.Core/Attributes/DiscordGatewayEventName.cs:29]: Shouldn't this class be documented? If only to make extending the library more accessible.

File: DSharpPlus.Test/Serialization/Core/TestPayloadSerialization.cs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+                foreach (string name in eventName.Names)
+                {
+                    payloadMap.Add(name, type);
+                }
+            }
+
+            JsonSerializer jsonSerializer = JsonSerializer.Create(new JsonSerializerSettings() { Converters = new List<JsonConverter>() { new DiscordSnowflakeConverter(), new OptionalConverter() } });
+            string filePath = "../../../../discord-payloads"; // This should be fine as long as we use the dotnet command instead of moving the executable and executing it.
+            foreach (string file in Directory.GetFiles(Path.Join(Environment.CurrentDirectory, filePath), "*.json", new EnumerationOptions() { RecurseSubdirectories = true }))
```
Comment (ExaInsanity) [inline DSharpPlus.Test/Serialization/Core/TestPayloadSerialization.cs:65]: i don't see why we're using Path.Join here - a relative path is just fine, plus to combine relative paths with absolute ones, especially with `..`s, we should be using `Path.GetFullPath("../../../../discord-payloads");`

**Review by ExaInsanity — COMMENTED** [review]
two nitpicks, other than that it looks good to me

note: i did not check back whether all gateway payload identifiers are correct, only for those on rest entities (because i thought it odd for them to be treated as payload objects).

File: DSharpPlus.Core/Attributes/DiscordGatewayEventName.cs (hunk 1 above)
Comment (OoLunar) [reply DSharpPlus.Core/Attributes/DiscordGatewayEventName.cs:29]: Oh absolutely, thanks for reminding me. I'll go through and document everything before merge

File: DSharpPlus.Test/Serialization/Core/TestPayloadSerialization.cs (hunk 2 above)
Comment (OoLunar) [reply DSharpPlus.Test/Serialization/Core/TestPayloadSerialization.cs:65]: Thanks for showing me this, I blame nothing but my own ignorance

**Pull request merged**